### PR TITLE
refactor(types): improve schema refinement

### PR DESCRIPTION
**Pull request by eventualbuddha**
zod is going through some restructuring, and v3 (in alpha) includes a feature that I really wanted before: custom dynamic error messages for refinements. It uses a new method on schemas called `superRefine` which may change, but for now is just what we need. It allows me to put the validation for various things, like enforcing unique ids, at the right place in the schema instead of at the root `Election`. This makes `safeParseElection` a pretty thin wrapper around the `Election` schema, whereas it wasn't before.

### CODE REVIEW

File: apps/bmd/src/AppCardlessVoting.test.tsx [hunk 1]
```diff
@@ -1,11 +1,8 @@
 import React from 'react'
 import { fireEvent, render, waitFor, within } from '@testing-library/react'
-import { asElectionDefinition } from '@votingworks/fixtures'
-import { parseElection } from '@votingworks/types'
 import * as GLOBALS from './config/globals'
 
-// import { electionSample } from '@votingworks/types'
-import electionSample from './data/electionSample.json'
+import { electionSampleDefinition } from './data'
```
Comment (eventualbuddha) [inline apps/bmd/src/AppCardlessVoting.test.tsx:5]: I didn't intend to make functional changes that would require anything outside `types` to change, but I made some minor changes that seem to have required updates in BMD. So here we are.

File: apps/bmd/src/components/__snapshots__/ElectionInfo.test.tsx.snap [hunk 2]
```diff
@@ -341,7 +341,7 @@ exports[`renders horizontal ElectionInfo with hash when specified 1`] = `
            
           <br />
           Election ID: 
-          66e55c0ffe
+          4aa260d7fc
```
Comment (eventualbuddha) [inline apps/bmd/src/components/__snapshots__/ElectionInfo.test.tsx.snap:344]: These now actually match the hash of the election.json files on disk. 🥳

File: apps/bmd/src/data/index.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import {
+  ElectionDefinition,
+  safeParseElectionDefinition,
+} from '@votingworks/types'
+import { readFileSync } from 'fs'
+import { join } from 'path'
+
+function electionDefinitionFromFile(
+  path: string
+): ElectionDefinition & { electionData: string } {
```
Comment (eventualbuddha) [inline apps/bmd/src/data/index.ts:10]: This was my shortcut for not updating `ElectionDefinition` to include `electionData` yet. I didn't want the scope creep in this PR.

File: apps/bmd/src/utils/Card.ts [hunk 4]
```diff
@@ -233,17 +233,17 @@ export class MemoryCard implements Card {
   /**
    * Inserts a simulated in-memory card with specified long and short values.
    */
-  public insertCard<T>(
+  public insertCard(
     shortValue?: string,
-    longValue?: T | string | Uint8Array
+    longValue?: string | Uint8Array
```
Comment (eventualbuddha) [inline apps/bmd/src/utils/Card.ts:238]: Allowing any arbitrary value and `JSON.stringify`ing it encouraged bad habits and made it hard/impossible to match election hashes properly.

File: libs/types/src/schema.ts [hunk 5]
```diff
@@ -26,50 +18,84 @@ export const Id = z
...
+      yield [index, id]
+    } else {
+      knownIds.add(id)
+    }
+  }
+}
 
 // Candidates
-export const Candidate = z.object({
+export const Candidate: z.ZodSchema<t.Candidate> = z.object({
```
Comment (eventualbuddha) [inline libs/types/src/schema.ts:47]: One of the main changes in this file is to explicitly annotate the types of the schemas. That helps us keep the types in sync (TS will complain if they diverge) and helps TS show better types for parsed values (i.e. `Election` instead of `{ title: string, … }`.

**Review by carolinemodic — APPROVED** [review]
